src/main/java/io/datanerds/avropatch/Patch.java

```diff
@@ -1,30 +1,40 @@
 package io.datanerds.avropatch;
 
+import io.datanerds.avropatch.operation.DefaultSchema;
 import io.datanerds.avropatch.operation.Operation;
+import io.datanerds.avropatch.value.conversion.*;
 import org.apache.avro.Schema;
 import org.apache.avro.io.*;
-import org.apache.avro.reflect.ReflectData;
-import org.apache.avro.reflect.ReflectDatumReader;
-import org.apache.avro.reflect.ReflectDatumWriter;
+import org.apache.avro.reflect.AvroSchema;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
 
+/**
+ * This class represents a JSON PATCH operation holding a sequence of operations to apply to a given object.
+ *
+ * @see <a href="https://tools.ietf.org/html/rfc6902">https://tools.ietf.org/html/rfc6902</a>
+ */
 public class Patch {
 
-    public static final Schema SCHEMA = ReflectData.get().getSchema(Patch.class);
-    private static final DatumWriter<Patch> writer = new ReflectDatumWriter<>(SCHEMA);
+    private static final PatchSerializer SERIALIZER = new PatchSerializer();
+
+    @AvroSchema(DefaultSchema.HEADERS)
+    private final Map<String, ?> headers;
     private final List<Operation> operations;
 
     public Patch() {
-        operations = new ArrayList<>();
+        this(new ArrayList<>(), new HashMap<>());
     }
 
     public Patch(List<Operation> operations) {
+        this(new ArrayList<>(operations), new HashMap<>());
+    }
+
+    public Patch(List<Operation> operations, Map<String, ?> headers) {
         this.operations = new ArrayList<>(operations);
+        this.headers = new HashMap<>(headers);
     }
 
     public boolean addOperation(Operation operation) {
```

Comment:
Thread safety will become an issue here. You are using a thread unsafe list for "operations". If there are 2 competing threads one of which does a getOperations and tries to iterate on it where as the other tries to add an operation at the same time, it will result in a ConcurrentModificationException. Would be better to use either a CopyOnWriteArrayList if operations are not expected to be added too frequently or synchronize getOperations and create a copy of the list before returning. AFAIK Collections.unmodifiableList is only a shallow copy and not a deep copy, so that would not handle it.